test/poolgirl_tests.erl

```diff
@@ -211,6 +214,33 @@ independent_childspec_pool() ->
     ok = supervisor:terminate_child(ExtSupervisor, PoolId),
     ok = supervisor:delete_child(ExtSupervisor, PoolId).
 
+graceful_internal_directory_concurrency() ->
+    _ = process_flag(trap_exit, true),
+    InternalDirectory = poolgirl_internal_directory:new(),
+    ?assertEqual(error, poolgirl_internal_directory:find(InternalDirectory, foobar)),
+
+    TestPid = self(),
+    Pid1 = spawn_link(
+             fun () ->
+                     ?assertEqual(true, poolgirl_internal_directory:register(InternalDirectory,
+                                                                             foobar)),
+                     TestPid ! proceed,
+                     receive stop -> ok end
+             end),
+    receive proceed -> ok end,
+    ?assertEqual({ok, Pid1}, poolgirl_internal_directory:find(InternalDirectory, foobar)),
+
+    % Unable to register - conflicting process is alive
+    ?assertEqual(false, poolgirl_internal_directory:register(InternalDirectory, foobar)),
+    Pid1 ! stop,
+    receive {'EXIT', Pid1, _} -> ok end,
+
+    % Able to register - conflicting process has terminated
+    ?assertEqual(true, poolgirl_internal_directory:register(InternalDirectory, foobar)),
+
+    % Able to re-register
+    ?assertEqual(true, poolgirl_internal_directory:register(InternalDirectory, foobar)).
```

Comment:
^ I was using `?assert` and `?assertNot` but it turns out they're entirely unhelpful when any of the asserted expressions returns something that's not a boolean (the assertion itself crashes...)